core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -85,7 +89,22 @@ public AllocationService(Settings settings, AllocationDeciders allocationDecider
         if (withReroute) {
             reroute(allocation);
         }
-        return new RoutingAllocation.Result(true, new RoutingTable.Builder().updateNodes(routingNodes).build().validateRaiseException(clusterState.metaData()));
+        RoutingTable routingTable = new RoutingTable.Builder().updateNodes(routingNodes).build().validateRaiseException(clusterState.metaData());
+        RoutingAllocation.Result result = new RoutingAllocation.Result(true, routingTable);
+
+        final int maxNumberOfLoggedShards = 10;
+        String startedShardsAsString = startedShards
+                .stream()
+                .limit(maxNumberOfLoggedShards)
+                .map(s -> String.format(Locale.ROOT, "[index: [%s], shard id: [%d]]", s.getIndex(), s.getId()))
+                .collect(Collectors.joining(", "));
+
+        logClusterHealthStateChange(
+                new ClusterStateHealth(clusterState),
+                new ClusterStateHealth(clusterState.metaData(), routingTable),
+                String.format(Locale.ROOT, "started %d shards: the first up to %d are [%s]", startedShards.size(), maxNumberOfLoggedShards, startedShardsAsString)
```

Comment:
we typically just use string concat. I think we can just do "shards started [{}]"..